File: 202-happy-number/happy-number.cpp
class Solution {
public:
    bool isHappy(int n) {
        if(n == 1 || n == 7) return true;
        set<int> seen;

        while(n){
            seen.insert(n);
            int sum = 0;
            while(n){
                int x = n%10;
                n = n/10;
                sum += x*x;
            }
            if(sum == 1 || sum == 7) return true;
            if(seen.find(sum) != seen.end()){
                return false;
            }

            n=sum;
        }

        return true;
    }
};

File: 202-happy-number/happy-number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        seen = set()
        if(n==1):
            return True
        while(n):
            seen.add(n)
            prod = 0
            
            while(n!=0):
                x = n%10
                n = n //10
                prod += x * x

            if(prod in seen):
                return False
            
            if(prod == 1):
                return True
            n = prod   

        return True